CONFIG.md
## Configuration

Outlined here are all of the configuration settings in this project.

`ticker`: string; ticker for CoinAPI - e.g LTC or BTC

`window`: number; the window of time (organized into four-hour steps) that the model considers at one time while predicting

`lookback`: number; amount of steps to consider for technical indicators

`model`: string; relative or absolute path to a model. Lightning .ckpt models are supported, as well as zipped Tensorflow/Keras .zip models

`coinapi-apikey`: string; API key for [CoinAPI](https://www.coinapi.io/)

`traintest-split`: float; split between training and testing data. Should be a float specifying the amount of testing data to use, out of 1. For example, 0.1 would indicate that 90% of training data should be used, and 10% should be used for testing.

`communication-options`: telegram options
- `telegram-api-key`: string; bot API key from Telegram's bot maker `BotFather`
- `telegram-password`: string; password used to authenticate with your Telegram bot

`trading-options`: live trading and backtest options
- `max-investment`: integer; maximum amount in coins/dollars to trade at one time (recommended to set to 100000000000 to effectively disable it, as setting max-trade is more effective)
- `max-trade`: float; % amount to trade at one time, as a float. For example, 0.2 would trade 20% of the funds in a certain side of the pair at a time
- `trading-fee`: float; trading fee per trade, out of 1. For example, 0.01 would be a 1% fee per trade.
- `poloniex-api-key`: string; API key for poloniex
- `poloniex-secret`: string; Poloniex secret key

`training-options`: options used for training
- `layer-neurons`: number; amount of neurons in a certain layer. Recommended values range from 50-250
- `layer-delta`: number; amount of neurons to decrease by, per layer. Recommended values range from 5-50
- `epochs`: number; amount of epochs to train for. This may not be reached because of early stopping when the validation loss is low enough.
- `batchsize`: number; amount of datapoints to pass to the model at once. Higher batchsizes make the model train faster, but can decrease financial performance.

`testing-options`: options for A/B testing
- `target`: string; any option in `training-options`, for example, `layer-neurons`
- `parameters`: array; parameters to manipulate the `target` with, as an array. For example, [100, 75, 50, 25] to test those values of `target`
- `graph`: boolean; whether to graph the coorelation between predicted and actual performance.
- `backtest`: boolean; whether to simulate training using the integrated backtest software

`tests`: experimental options
- `lightning`: boolean; whether to enable PyTorch Lightning. This is only useful for training models, as backtesting and trading are not implemented yet.

src/config.py
import yaml
import os


def get_config():
    # OPTIONS
    ticker = "LTC"
    window = 14
    lookback = 14
    model = (
        None  # if Tensorflow export zipfile is set, it will be used instead of training
    )
    coinapi_apikey = ""
    traintest_split = 0.2

    # TRADING OPTIONS
    max_investment = 1000000  # basically disabled
    max_trade = 0.4  # 40% of the balance
    # trading_fee = 0.001015 # 0.25%
    trading_fee = 0.002
    poloniex_api_key = ""
    poloniex_secret = ""

    # TELEGRAM BOT OPTIONS
    telegram_api_key = ""
    telegram_password = ""

    # TRAINING OPTIONS
    layer_neurons = 175
    layer_delta = 25
    epochs = 100
    batchsize = 64

    # TESTING OPTIONS
    target = ""
    parameters = []
    opt_graph = False
    opt_backtest = False
    
    # MORE TESTING OPTIONS
    lightning = False

    # test if config.yaml exists

    if os.path.exists("config.yaml"):
        with open("config.yaml", "r") as file:
            config = yaml.safe_load(file)
            ticker = config["ticker"]
            window = config["window"]
            lookback = config["lookback"]
            model = config["model"]
            coinapi_apikey = config["coinapi-apikey"]
            traintest_split = config["traintest-split"]

            max_investment = config["trading-options"]["max-investment"]
            max_trade = config["trading-options"]["max-trade"]
            trading_fee = config["trading-options"]["trading-fee"]
            poloniex_api_key = config["trading-options"]["poloniex-api-key"]
            poloniex_secret = config["trading-options"]["poloniex-secret"]

            telegram_api_key = config["communication-options"]["telegram-api-key"]
            telegram_password = config["communication-options"]["telegram-password"]

            telegram_api_key = config["communication-options"]["telegram-api-key"]
            telegram_password = config["communication-options"]["telegram-password"]

            layer_neurons = config["training-options"]["layer-neurons"]
            layer_delta = config["training-options"]["layer-delta"]
            epochs = config["training-options"]["epochs"]
            batchsize = config["training-options"]["batchsize"]

            target = config["testing-options"]["target"]
            parameters = config["testing-options"]["parameters"]
            opt_graph = config["testing-options"]["graph"]
            opt_backtest = config["testing-options"]["backtest"]
            
            tests = {
                "lightning": config["tests"]["lightning"]
            }

            return (
                ticker,
                window,
                lookback,
                model,
                coinapi_apikey,
                traintest_split,
                max_investment,
                max_trade,
                trading_fee,
                poloniex_api_key,
                poloniex_secret,
                telegram_api_key,
                telegram_password,
                layer_neurons,
                layer_delta,
                epochs,
                batchsize,
                target,
                opt_graph,
                opt_backtest,
                parameters,
                tests
            )
    else:
        print("config.yaml not found, using defaults")
        config = {
            "ticker": ticker,
            "window": window,
            "lookback": lookback,
            "model": model,
            "coinapi-apikey": coinapi_apikey,
            "traintest-split": traintest_split,
            "trading-options": {
                "max-investment": max_investment,
                "max-trade": max_trade,
                "trading-fee": trading_fee,
                "poloniex-api-key": poloniex_api_key,
                "poloniex-secret": poloniex_secret,
            },
            "communication-options": {
                "telegram-api-key": telegram_api_key,
                "telegram-password": telegram_password,
            },
            "training-options": {
                "layer-neurons": layer_neurons,
                "layer-delta": layer_delta,
                "epochs": epochs,
                "batchsize": batchsize,
            },
            "testing-options": {
                "target": target,
                "parameters": parameters,
                "graph": opt_graph,
                "backtest": opt_backtest,
            },
            "tests": {
                "lightning": lightning
            }
        }

        yaml.dump(config, open("config.yaml", "w"))

        print(
            f"config.yaml created with default values. Please fill in the required fields and run again."
        )
        exit(0)
